### ISSUE TITLE
twisted.internet.kqueuereactor doesn't inspect an exception's errno properly on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7887]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7887|
|Type|defect|
|Created|2015-05-07 09:13:26Z|
|Branch|https://github.com/twisted/twisted/tree/kqueue-eintr-7887-3|

The offending code:

```
        except OSError as e:
            if e[0] == errno.EINTR:
                return
            else:
                raise
```

This should do ``` e.errno == errno.EINTR ``` instead. There's no test for this, it seems, so one should be added too.

<details><summary>Searchable metadata</summary>

```
trac-id__7887 7887
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_kqueue_eintr_7887_3 branches/kqueue-eintr-7887-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1430990006153619 1430990006153619
changetime__1438310834300981 1438310834300981
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44785]) Branching to kqueue-eintr-7887.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Added a failing test, fixed the functionality for it to pass, passes. Builders spun, look green so far.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks!  I take it this is part of the py3 port for kqreactor?

Some things to address:

    1. `_IKQueueProvider` is a great idea. More explicit interfaces are good. That said, interfaces shouldn't have "provider" in their name.  This is just `_IKQueueCreator` or something.
    2. `_StdlibKQueue` is unnecessary.  The provider of the interface as declared by `_IKQueueProvider` is the `select` module itself, which already has all the attributes that you need.  Rather than tediously copying them, you can simply call `zope.interface.declarations.directlyProvides(select, _IKQueueProvider)` right after declaring it.
    3. I was a little suspicious about lines 61 and 61 of `kqreactor.py`, and ... yeah, I don't think test coverage is sufficient:
```
>>> from twisted.internet.kqreactor import _StdlibKQueue
>>> skq = _StdlibKQueue()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File ".../twisted/internet/kqreactor.py", line 61, in __init__
    self.KQ_FILTER_READ, self.KQ_FILTER_WRITE, self.KQ_EV_DELETE,
AttributeError: '_StdlibKQueue' object has no attribute 'KQ_FILTER_READ'
```
    4. the documentation for the `_kqueueImpl` argument to `KQueueReactor.__init__` is wrong.  It's a 0-argument factory.
    5. The documentation for `_IKQueueProvider` is wrong.  First, `@param` is not how an attribute is documented; that would be `@ivar` normally, but on an interface it should be an `Attribute` instance.  Also, none of the attributes (`KQ_FILTER_READ` etc) are documented.

and some suggestions:

    1. `_kqueueImpl` seems redundant.  I'd just call it `_impl`.
    2. The indirection of the constants (`KQ_FILTER_*` and `KQ_EV_*`) doesn't seem super helpful to me.  I can see how it would be useful in a test to see a `twisted.python.constants`-style `repr` rather than an integer, but it's a lot of indirection for a very small amount of test utility and zero implementation utility.  Instead, it seems to me like you could just pass in the something that can create `self._kq` and a `kevent`, which would be a lot simpler implementation-wise (no need to indirect anywhere but `KQueueReactor.__init__` and `.beforeDaemonize`).
    3. I feel like this error is repeated in other reactors.  Specifically, I can see that this ''exact'' issue (comparison of `args[0]` to `EINTR`) repeated in `twisted.internet.pollreactor` and `twisted.trial._dist.workertrial`.  Do you have any thoughts for how we might catch and fix more of these?  Any general patterns for testing?  Or will each case need to be a one-off?

I think given the test-coverage problem, this one needs another trip through re-review.  Looking forward to it!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl removed owner|
|-|-|

Hi glyph, thanks for the review.

I've tried to address the comments -- I think it's in a bit of a better shape. I removed ```_StdlibKQueue```, and renamed the IKQueueProvider. 

I ran it locally, and on OS X this test passes, and on OS X with KQueue enabled in reactorbuilder, kqreactor.py gets 86%. We really need to get a KQueue platform supported on Py3.

I also took the opportunity to fix up some twistedchecker errors -- namely some inconsistent spacing and missing newlines.

Builders spun, looking green, please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [44932]) Branching to kqueue-eintr-7887-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

oops i should put the tag on this hey

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

We now have a FreeBSD 10.1 + Py3.4 builder. Please see the test run for this unsupported builder https://buildbot.twistedmatrix.com/builders/freebsd10.1-py3.4/builds/7 (green) vs the same builder on trunk https://buildbot.twistedmatrix.com/builders/freebsd10.1-py3.4/builds/6 (red).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this.

I find the whole test and _IKQueue name confusing.... same comment for _kq and _impl members.

Maybe we can renamed the interface to _IKEventNotification ...  and _impl to _kEventNotification ...and _kq to _activeKQueue

My suggestions are based on the info from the man page

```
 kqueue, kevent -- kernel event notification mechanism
```


--------

```
L{KQueueReactor} should handle L{errno.EINTR} by returning.
```

I feel that this docstring is incomplete.... as including 'should' in the test docstring does not add to much info.

Python docs are not very helpful and for kqueue.control I see `Low level interface to kevent` ... and I have no idea how and when should I call this method.



-------

Here is a quick untested refactoring of the test .... with a few inline comments

For now we have a single test and I find it easier to read if all test dependencies are inside the test definition.

... assertIsNone(x) is only available in python2.7 ... but maybe we can implement it in Trial

```
# Copyright (c) Twisted Matrix Laboratories.
# See LICENSE for details.

"""
Tests for L{twisted.internet.kqueuereactor}.
"""

from __future__ import division, absolute_import
import errno

from zope.interface import implementer

from twisted.trial.unittest import TestCase

try:
    from twisted.internet.kqreactor import KQueueReactor, _IKEventNotification
    kqueueSkip = None
except ImportError:
    kqueueSkip = "KQueue not available."


class KQueueTests(TestCase):
    """
    These are tests for L{KQueueReactor}'s implementation, not its real world
    behaviour. For that, look at
    L{twisted.internet.test.reactormixins.ReactorBuilder}.
    """
    skip = kqueueSkip

    def test_EINTR(self):
        """
        L{KQueueReactor} handles L{errno.EINTR} by returning ... WHAT?
        """

        @implementer(_IKEventNotification)
        class FakeKEventNotification(object):
            """
            Event notification mechanism using rigged low level implementations.
            """
            kqueue = LowLevelKQueue
            kevent = LowLevelKEvent

        class LowLevelKQueue(object):
            """
            KQueue implementation that raises L{errno.EINTR} when C{control} is called.
            """
            def control(self, *args, **kwargs):
                raise OSError(errno.EINTR, "Interrupted")

        def LowLevelKEvent(*args, **kwargs):
            """
            Do nothing.
            """
            pass

        reactor = KQueueReactor(
         _kEventNotification=FakeKEventNotification()))

        # WHY care for result?
        # from code and docs doKEvent does not return any value http://twistedmatrix.com/documents/current/api/twisted.internet.kqreactor.KQueueReactor.html#doKEvent
        # How is the code actioned by pooling for events without a timeout?
        result = reactor.doKEvent(0)

        self.assertEqual(None, result)

```


-----------

overall the changes are good and the fix is valid :) 

I think that the test should be updated to make it easier to follow what it want to check and how it does that.

Otherwise it could prove to be hard for future contributors to refactor this code.

Please review my comments. If you feel that the code is already well documented, feel free to resubmit it again for review so that it could be checked by another dev.

Many thanks for working on this!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl removed owner|
|-|-|

Hmm. I don't agree with all the comments, but I did add some more documentation about what it does and why, as well as fix up the test a little.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @hawkowl|
|-|-|

https://buildbot.twistedmatrix.com/builders/python-3k-warnings/builds/818/steps/trial/logs/problems `twisted.scripts.test.test_scripts.Tap2DeprecationTests.test_tap2debDeprecation` is failing, of all things, hopefully this is unrelated (I re-spun the builders).

Some nitpicking:

  1. I'm wondering if we should even bother talking about a 5-year-old maintenance release of a 7-year-old version of Python in the documentation here.
  2. You appear to be overly-aggressively wrapping the `_updateRegistration` lines; by my count they were only 78 columns previously.
  3. I don't find the comment ("`a changelist (which is always none)`") terribly illuminating.  I'd be inclined to just delete it.
  4. `EINTR` can happen for lots of reasons besides `^C`; the comment should not make it sound as though this is the only one.  (In fact, given that we use `signalfd` and `SA_RESTART` we should _not_ usually get `EINTR` in that case; the fact that we are is weird.  In what circumstances has it been observed?)
  5. Thanks for fixing the 3-space indents.  Shudder.
  6. Don't ever use `pass`.  In this case it's entirely unnecessary; the docstring already holds the place of the function body for `_fakeKevent`.
  7. `makeFakeKQueue` really needs some `@param`s and a `@return`.

Please address these to your satisfaction, and, assuming the buildbots remain happy, land.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

(In [45313]) Branching to kqueue-eintr-7887-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

Thanks for the review, Glyph.

- The tap2deprecation test was fixed in trunk, merged forward to verify.
- Removed.
- Fixed.
- Removed.
- Altered the comment. It's happened in the tests, so, I'm not sure what triggers it.
- :)
- Fixed.
- Fixed.

Buildbot looks happy, the FreeBSD 10.1 Py3.4 builder passes now. Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45317]) Merge kqueue-eintr-7887-3: Fix twisted.internet.kqueuereactor doesn'tinspect an exception's errno properly on Python 3

Author: hawkowl
Reviewers: adiroiban, glyph
Fixes: [#7887](https://github.com/twisted/twisted/issues/7887)
